File: static/js/Item.js
class Item {

  constructor (id, title, location, image, properties, x, y, interact_x, interact_y,
               lookID, smellID, tasteID, takeID, touchID, speakID) {
    var self = this;
    
    const OBJ_URL = 'img/objects/';
    this.OBJ_URL = OBJ_URL;
    
    this.id = id;
    this.title = title;
    this.location = location;
    this.height;
    this.width;
    this.x = x;
    this.y = y;
    this.image = image;
    this.properties = properties;
    this.image_opened = properties.image_opened;
    this.image_closed = properties.image_closed;
    this.takeable = properties.takeable;
    this.interaction_x = interact_x;
    this.interaction_y = interact_y;

    this.inventory_open = false;
    
    this.lookID = lookID;
    this.smellID = smellID;
    this.tasteID = tasteID;
    this.takeID = takeID;
    this.touchID = touchID;
    this.speakID = speakID;
    
    this.lookText;
    this.smellText;
    this.tasteText;
    this.takeText;
    this.touchText;
    this.speakText;

    if (this.properties.is_container) {
      if (this.properties.is_open == 0) {
        var imgURL = OBJ_URL + this.image_closed;
      } else {
        var imgURL = OBJ_URL + this.image_opened;
      }
    } else {
      var imgURL = OBJ_URL + this.image;
    }

    this.img = $('<img data-objid="' + this.id + '" class="object">');
    this.img[0].onload = function() {
      self.width = $(self.img).width();
      self.height = $(self.img).height();
    }
    $(this.img).attr('src', imgURL);
    $(this.img).css('left', this.x);
    $(this.img).css('top', this.y);
    $('main').append(this.img);

    $(this.img).click(function(event) {
      event.preventDefault();
      event.stopPropagation();
      if (!is_ui_disabled) {
        showMenu(this, event.pageX - $('main').offset().left, event.pageY - $('main').offset().top);
      }
    });
  }

  openIfClosed() {
    var self = this;
    $.getJSON(SERVICE_URL + 'item/' + self.id + '/properties', function(data) {
      if (!data.is_open && !data.is_locked) {
        $(self.img).attr('src', self.OBJ_URL + self.image_opened);
        $.ajax({
          url: SERVICE_URL + 'item/' + self.id + '/open',
          type: 'PUT'
        });
      }
    });
  }
  closeIfOpened() {
    console.log('close');
    console.log(this.id);
    var self = this;
    $.getJSON(SERVICE_URL + 'item/' + self.id + '/properties', function(data) {
      if (data.is_open) {
        $(self.img).attr('src', self.OBJ_URL + self.image_closed);
        $.ajax({
          url: SERVICE_URL + 'item/' + self.id + '/close',
          type: 'PUT'
        });
      }
    });
  }
  openInventory() {
    var self = this;
    this.inventory_open = true;
    $.getJSON(SERVICE_URL + 'player/' + pc.id + '/inventory', function(inventory) {
      $.get(TEMPLATE_URL + 'item_inventory.html', function(template) {
        template = $(template);
        var img = $('<img src="' + self.OBJ_URL + self.image_opened + '">');
        $(template).find('.objContents > figure').append(img);
        $(template).find('.objContents > figure').append('<figcaption>' + ucwords(self.title) + '</figcaption>');
        $(template).find('header > i').click(function(event) {
          event.preventDefault();
          event.stopPropagation();
          self.closeInventory();
        });
        $('main').append(template);
        $(inventory).each(function(index, item) {
          var div = $('<div class="item_container" data-objid="' + item.oid + '"><img src="' + self.OBJ_URL + item.image_closed + '" alt="' + item.title + '" title="' + item.title + '"></div>');
          $($('.playerInventory').find('td')[index]).html(div);
        });
        $('.playerInventory td').each(function(index, item) {
          $(item).droppable({
            drop: function(event, ui) {
              var cloned = $(ui.draggable).clone();
              $(this).append(cloned);
              $(cloned).draggable({
                containment: '#object_inventory .fg > div',
                helper: 'clone',
                appendTo: '#object_inventory .fg > div',
                start: function(event, ui) {
                  $(this).remove();
                }
              });
              $.post(SERVICE_URL + 'item/' + cloned.attr('data-objid') + '/take/' + self.id, function(data) {
                console.log(data);
              });
            }
          });
          $(item).children('div').draggable({
            containment: '#object_inventory .fg > div',
            helper: 'clone',
            appendTo: '#object_inventory .fg > div',
            start: function(event, ui) {
              $(this).remove();
            }
          });
        });
        $.getJSON(SERVICE_URL + 'item/' + self.id + '/inventory', function(data) {
          $(data).each(function(index, item) {
            var div = $('<div class="item_container" data-objid="' + item.oid + '"><img src="' + self.OBJ_URL + item.image_closed + '" alt="' + item.title + '" title="' + item.title + '"></div>');
            $($('.objectItems').find('td')[index]).html(div);
            $(div).draggable({
              containment: '#object_inventory .fg > div',
              helper: 'clone',
              appendTo: '#object_inventory .fg > div',
              start: function(event, ui) {
                $(this).remove();
              }
            });
          });
          $('.objectItems').find('td').droppable({
            drop: function(event, ui) {
              var cloned = $(ui.draggable).clone();
              $(this).append(cloned);
              $(cloned).draggable({
                containment: '#object_inventory .fg > div',
                helper: 'clone',
                appendTo: '#object_inventory .fg > div',
                start: function(event, ui) {
                  $(this).remove();
                }
              });
              $.ajax({
                url: SERVICE_URL + 'item/' + cloned.attr('data-objid') + '/drop/' + self.id,
                type: 'PUT',
                success: function(result) {
                    // Do something with the result
                }
              });
            }
          });
        });
      });
    });
  }
  
  closeInventory() {
    $('#object_inventory').remove();
    this.inventory_open = false;
  }
}

File: Item/Item.go
package Item

import (
  "log"
  "github.com/jonathanface/bumquest/DBUtils"
)

type Item struct {
  Oid int `json:"oid"`
  Title string `json:"title"`
  Image string `json:"image"`
  Is_takeable int `json:"is_takeable"`
  X int `json:"x"`
  Y int `json:"y"`
  Interact_x int `json:"interaction_x"`
  Interact_y int `json:"interaction_y"`
  Look_id int `json:"look_id"`
  Taste_id int `json:"taste_id"`
  Smell_id int `json:"smell_id"`
  Take_id int `json:"take_id"`
  Touch_id int `json:"touch_id"`
  Speak_id int `json:"speak_id"`
  Properties Info `json:"properties"`
}

type Info struct {
  Description string `json:"description"`
  Is_open int `json:"is_open"`
  Is_locked int `json:"is_locked"`
  Is_takeable int `json:"is_takeable"`
  Is_container int `json:"is_container"`
  Image_opened string `json:"image_opened"`
  Image_closed string `json:"image_closed"`
}

func GetProperties(id int) Info {
  var info = Info{}
  db := DBUtils.OpenDB()
  db.QueryRow("SELECT is_open,is_locked,is_takeable,is_container,image_closed,image_opened FROM object_properties WHERE objectID = ?", id).Scan(&info.Is_open, &info.Is_locked, &info.Is_takeable, &info.Is_container, &info.Image_closed, &info.Image_opened)
  DBUtils.CloseDB(db)
  return info
}

func GetActionText(table string, id int) Info {
  var info = Info{}
  db := DBUtils.OpenDB()
  db.QueryRow("SELECT text FROM " + table + " WHERE objectID = ?", id).Scan(&info.Description)
  DBUtils.CloseDB(db)
  return info
}


func GetInventory(oid int) []Item {
  db := DBUtils.OpenDB();
  rows, err := db.Query("SELECT objectID FROM object_inventory WHERE containerID = ?", oid)
  if (err != nil) {
    log.Fatal(err)
  }
  
  var items []Item
  for rows.Next() {
    var currentOid int
    rows.Scan(&currentOid)
    log.Println(currentOid)
    objRow := db.QueryRow("SELECT objectID,name,image,x,y,interact_x,interact_y," + 
                           "lookID, takeID, smellID, tasteID, touchID, speakID FROM objects WHERE objectID = ?", currentOid)
    item := Item{}
    err = objRow.Scan(&item.Oid,&item.Title,&item.Image,&item.X,&item.Y,&item.Interact_x, &item.Interact_y,
                      &item.Look_id, &item.Take_id, &item.Smell_id, &item.Taste_id, &item.Touch_id, &item.Speak_id)
    if err != nil {
      log.Fatal(err)
    }
    items = append(items, item)
  }
  DBUtils.CloseDB(db)
  return items
}